app/plugin_install.go

```diff
@@ -413,21 +419,81 @@ func (a *App) removePlugin(id string) *model.AppError {
 	if err = a.removeSignature(id); err != nil {
 		mlog.Error("Can't remove signature", mlog.Err(err))
 	}
-
-	a.notifyClusterPluginEvent(
-		model.CLUSTER_EVENT_REMOVE_PLUGIN,
-		model.PluginEventData{
-			Id: id,
-		},
-	)
+	canaryTag := *a.Config().PluginSettings.CanaryTag
+	if canaryTag == "" {
+		a.notifyClusterPluginEvent(
+			model.CLUSTER_EVENT_REMOVE_PLUGIN,
+			model.PluginEventData{
+				Id: id,
+			},
+		)
+	}
 
 	if err := a.notifyPluginStatusesChanged(); err != nil {
 		mlog.Error("Failed to notify plugin status changed", mlog.Err(err))
 	}
 
+	if canaryTag == "" {
+		a.removeAllPlugins(id)
+	} else {
+		a.installPluginWithNoCanaryTag(id, pluginEnabled)
+	}
+
 	return nil
 }
 
+func (a *App) installPluginWithNoCanaryTag(id string, pluginEnabled bool) {
+	pluginPath := filepath.Join(fileStorePluginFolder, fmt.Sprintf("%s.tar.gz", id))
+	signaturePath := a.getSignatureStorePath(id)
+	bundleExist, appErr := a.FileExists(pluginPath)
+	if appErr != nil {
+		mlog.Warn("No plugin without canary tag", mlog.Err(appErr))
+		return
+	}
+	if !bundleExist {
+		mlog.Warn("No plugin without canary tag")
+		return
+	}
+	pluginReader, err := os.Open(pluginPath)
+	if err != nil {
+		mlog.Error("Failed to open plugin without canary tag", mlog.Err(err), mlog.String("path", pluginPath))
+		return
+	}
+	defer pluginReader.Close()
+	signatureReader, err := os.Open(signaturePath)
+	if err != nil {
+		mlog.Info("No signature file for plugin without canary tag", mlog.Err(err), mlog.String("path", pluginPath))
+	} else {
+		defer signatureReader.Close()
+	}
+
+	_, appErr = a.installPluginLocally(pluginReader, signatureReader, installPluginLocallyAlways)
+	if appErr != nil {
+		mlog.Error("Unable to install plugin without cannary tag", mlog.Err(appErr), mlog.String("path", pluginPath))
+	}
+	if pluginEnabled {
+		appErr = a.EnablePlugin(id)
+		if appErr != nil {
+			mlog.Error("Unable to enable plugin without cannary tag", mlog.Err(appErr), mlog.String("path", pluginPath))
+		}
+	}
+}
+
+func (a *App) removeAllPlugins(id string) {
+	fileStorePaths, appErr := a.ListDirectory(fileStorePluginFolder)
+	if appErr != nil {
+		mlog.Error("Can't list directory", mlog.Err(appErr), mlog.String("path", fileStorePluginFolder))
+		return
+	}
+	for _, fileStorePath := range fileStorePaths {
+		if strings.Contains(fileStorePath, id) {
```

Comment:
Some of our plugin IDs are single word, e.g. `zoom` or `jira`. If another plugin happened to contain that as a substring, it seems like it would be incorrectly removed here. Can we parse the file store path and examine the id discretely?